libnm/nm-device-wireguard.c

```diff
@@ -0,0 +1,254 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: t; c-basic-offset: 4 -*- */
+/*
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ * Copyright 2018 Javier Arteaga <jarteaga@jbeta.is>
+ */
+
+#include "nm-default.h"
+
+#include "nm-device-wireguard.h"
+#include "nm-object-private.h"
+
+G_DEFINE_TYPE (NMDeviceWireguard, nm_device_wireguard, NM_TYPE_DEVICE)
+
+#define NM_DEVICE_WIREGUARD_GET_PRIVATE(o) (G_TYPE_INSTANCE_GET_PRIVATE ((o), NM_TYPE_DEVICE_WIREGUARD, NMDeviceWireguardPrivate))
+
+typedef struct {
+	guchar *private_key;
+	guchar *public_key;
+	guint16 listen_port;
+	guint32 fwmark;
+} NMDeviceWireguardPrivate;
+
+NM_GOBJECT_PROPERTIES_DEFINE_BASE (
+	PROP_PRIVATE_KEY,
+	PROP_PUBLIC_KEY,
+	PROP_LISTEN_PORT,
+	PROP_FWMARK,
+);
+
+/**
+ * nm_device_wireguard_get_private_key:
+ * @device: a #NMDeviceWireguard
+ *
+ * Gets the private key for this interface
+ *
+ * Returns: pointer to the 32-byte private key
+ *
+ * Since: 1.14
+ **/
+const guchar*
+nm_device_wireguard_get_private_key (NMDeviceWireguard *device)
+{
+	g_return_val_if_fail (NM_IS_DEVICE_WIREGUARD (device), 0);
+
+	return NM_DEVICE_WIREGUARD_GET_PRIVATE (device)->private_key;
+}
+
+/**
+ * nm_device_wireguard_get_public_key:
+ * @device: a #NMDeviceWireguard
+ *
+ * Gets the public key for this interface
+ *
+ * Returns: pointer to the 32-byte public key
+ *
+ * Since: 1.14
+ **/
+const guchar*
+nm_device_wireguard_get_public_key (NMDeviceWireguard *device)
+{
+	g_return_val_if_fail (NM_IS_DEVICE_WIREGUARD (device), 0);
+
+	return NM_DEVICE_WIREGUARD_GET_PRIVATE (device)->public_key;
+}
+
+/**
+ * nm_device_wireguard_get_listen_port:
+ * @device: a #NMDeviceWireguard
+ *
+ * Gets the local UDP port this interface listens on
+ *
+ * Returns: UDP listen port
+ *
+ * Since: 1.14
+ **/
+guint
+nm_device_wireguard_get_listen_port (NMDeviceWireguard *device)
+{
+	g_return_val_if_fail (NM_IS_DEVICE_WIREGUARD (device), 0);
+
+	return NM_DEVICE_WIREGUARD_GET_PRIVATE (device)->listen_port;
+}
+
+/**
+ * nm_device_wireguard_get_fwmark:
+ * @device: a #NMDeviceWireguard
+ *
+ * Gets the firewall mark (fwmark) byte, if this interface sets one
+ *
+ * Returns: 0 if fwmark not in use, fwmark byte value otherwise
+ *
+ * Since: 1.14
+ **/
+guint
```

Comment:
guchar, if it's really a byte?

Reply:
Gah - no, it's not a byte, it's 32-bit. The explanation is wrong (in a few places too).